src/main/java/frc/robot/subsystems/vision/PhotonVisionIO.java

```diff
@@ -0,0 +1,90 @@
+    private final PhotonPoseEstimator estimator;
+    private Result result;
+
+    public PhotonVisionIO(PhotonCamera camera, Transform3d robotToCamera) {
+        this.camera = camera;
+
+        try {
+            estimator = new PhotonPoseEstimator(
+                    AprilTagFields.k2023ChargedUp.loadAprilTagLayoutField(),
+//                    PhotonPoseEstimator.PoseStrategy.MULTI_TAG_PNP,
+                    PhotonPoseEstimator.PoseStrategy.AVERAGE_BEST_TARGETS,
+                    camera,
+                    robotToCamera
+            );
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void setPipeLine(int pipeLineIndex) {
+        camera.setPipelineIndex(pipeLineIndex);
+    }
+
+    @Override
+    public void updateInputs(VisionInputs inputs) {
+        var latestResult = camera.getLatestResult();
+        for (int i = 0; i < latestResult.targets.size(); i++) {
+        System.out.println(latestResult.targets.get(i));
+        }
+
+        if (latestResult != null) {
+            inputs.latency = (long) latestResult.getLatencyMillis();
+            if (latestResult.getBestTarget() != null) {
+                inputs.area = latestResult.getBestTarget().getArea();
+                inputs.pitch = latestResult.getBestTarget().getPitch();
+                inputs.yaw = latestResult.getBestTarget().getYaw();
+                inputs.targetSkew = latestResult.getBestTarget().getSkew();
+                inputs.hasTargets = latestResult.hasTargets();
+                inputs.targetID = latestResult.getBestTarget().getFiducialId();
+            }
+
+            var estimatedPose = estimator.update(latestResult);
+            if (estimatedPose.isPresent()) {
+                var pose = estimatedPose.get().estimatedPose;
+                inputs.poseFieldOriented = new double[]{
+                        pose.getX(),
+                        pose.getY(),
+                        pose.getZ(),
+                        pose.getRotation().getX(),
+                        pose.getRotation().getY(),
+                        pose.getRotation().getZ()
+                };
+
+                var cameraToTarget = latestResult.getBestTarget().getBestCameraToTarget();
+                System.out.println("cameraToTarget: " + cameraToTarget);
+                inputs.cameraToTarget = new double[]{
+                        cameraToTarget.getX(),
+                        cameraToTarget.getY(),
+                        cameraToTarget.getZ(),
+                        cameraToTarget.getRotation().getX(),
+                        cameraToTarget.getRotation().getY(),
+                        cameraToTarget.getRotation().getZ()
+                };
+
+                result = new Result(latestResult.getTimestampSeconds(), estimatedPose.get().estimatedPose);
+            } else {
+                result = null;
+            }
+        } else {
+            result = null;
+        }
+    }
```

Comment:
It looks like `latestResult` will never be `null`. But:
1. You don't `null`-check in the for loop above.
2. Even if you do need to `null`-check, it's better to check whether `latestResult == null` and then immediately `return` because it will help you reduce indentation, which helps you reduce indentation. The same can be done with the condition in line 55:
```java
        var latestResult = camera.getLatestResult();
        for (int i = 0; i < latestResult.targets.size(); i++) {
            System.out.println(latestResult.targets.get(i));
        }

        if (latestResult == null) {
            result = null;
            return;
        }
        inputs.latency = (long) latestResult.getLatencyMillis();
        if (latestResult.getBestTarget() == null) {
            return;
        }
        inputs.area = latestResult.getBestTarget().getArea();
        inputs.pitch = latestResult.getBestTarget().getPitch();
        inputs.yaw = latestResult.getBestTarget().getYaw();
        inputs.targetSkew = latestResult.getBestTarget().getSkew();
        inputs.hasTargets = latestResult.hasTargets();
        inputs.targetID = latestResult.getBestTarget().getFiducialId();

        var estimatedPose = estimator.update(latestResult);
        if (estimatedPose.isEmpty()) {
            result = null;
            return;
        }
        var pose = estimatedPose.get().estimatedPose;
        inputs.poseFieldOriented = new double[]{
                pose.getX(),
                pose.getY(),
                pose.getZ(),
                pose.getRotation().getX(),
                pose.getRotation().getY(),
                pose.getRotation().getZ()
        };

        var cameraToTarget = latestResult.getBestTarget().getBestCameraToTarget();
        System.out.println("cameraToTarget: " + cameraToTarget);
        inputs.cameraToTarget = new double[]{
                cameraToTarget.getX(),
                cameraToTarget.getY(),
                cameraToTarget.getZ(),
                cameraToTarget.getRotation().getX(),
                cameraToTarget.getRotation().getY(),
                cameraToTarget.getRotation().getZ()
        };

        result = new Result(latestResult.getTimestampSeconds(), estimatedPose.get().estimatedPose);
```